Modify rule S4335: Adapt to LaYC (#2693)
SonarSource/rspec: rules/S4335/javascript/rule.adoc

== Why is this an issue?

> In TypeScript, intersections are used to combine multiple types into a single one, allowing you to create more flexible and powerful type definitions. An intersection type is represented using the ampersand symbol ``++&++``. It allows you to combine multiple types into a single type that includes all the properties and methods from each type.

Reviewer: ```suggestion
In TypeScript, intersections are used to combine multiple types into a single one. An intersection type is represented using the ampersand symbol ``++&++``. It allows you to combine multiple types into a single type that includes all the properties and methods from each type, thus creating more flexible and powerful type definitions.
```

Reviewer: Let's avoid saying what it "allows" twice :)

However, some of the basic types of TypeScript should not be used with intersections:

* The `never` type represents the type of values that never occur. Intersecting any type with `never` will always result in type `never`.

> * The `any` type allows to opt-out of type checking during compilation. Variables of type any allow you to access arbitrary properties, even ones that don't exist. `any` comes at the cost of losing type safety, which is one of the main motivations for using TypeScript. Avoid using `any` when not necessary. Intersecting any type with `any` will always result in type `any`.

Reviewer: ```suggestion
* The `any` type allows to opt-out of type checking during compilation. Expressions of type `any` allow you to access arbitrary properties, even ones that don't exist. `any` comes at the cost of losing type safety, which is one of the main motivations for using TypeScript. Avoid using `any` when not necessary. Intersecting any type with `any` will always result in type `any`.
```

> * The `undefined` and `null` are the types for their respective values. Intersecting any type with `never` will always result in type `never`.

Reviewer: ```suggestion
* The `undefined` and `null` types are the types for their respective value. Intersecting any type with `never` will always result in type `never`.
```

Reviewer: Something is weird with the second sentence. Probably copy-pasted :/

> * The `void` type implies the absence of a type. Intersecting any type with `never` will always result in type `never`.

Reviewer: Same here.

Additionally, an intersection with a type without members (for example, ``++{}++``) doesn't change the resulting type, is redundant, and can be safely removed from the intersection.

[source,javascript,diff-id=1,diff-type=noncompliant]
----

> type MyType = MyOtherType & null; //Noncompliant: MyType is always 'never'

Reviewer: ```suggestion
type Foo = T & null; // Noncompliant: 'Foo' is always 'never'
```

> type MyType = MyOtherType & any; //Noncompliant: MyType is always 'any'

Reviewer: ```suggestion
type Bar = T & any; // Noncompliant: 'Bar' is always 'any'
```

> type MyType = MyOtherType & AnotherType & {}; //Noncompliant: '{}' has no memebers and is redundant

Reviewer: ```suggestion
type Baz = T & U & {}; // Noncompliant: '{}' has no members and is redundant
```

----

> Use specific types that accurately represent the data being used.

Reviewer: ```suggestion
Use consistent types that accurately reflect and restrict the domain of values of the defined data type.
```

Author: I would not say restrict, intersections in principle are used to extend, not restrict, right?

Reviewer: Yeah, but if you use `any`, you kinda open the gate to everything, right ?

Reviewer: Maybe restrict is too strong. What about limit or narrow?

Author: but intersecting with any raises an issue. I changed the compliant version.

what about 

```suggestion
Use consistent types that accurately reflect the domain of values of the defined data type.
```

Reviewer: Looks good

[source,javascript,diff-id=1,diff-type=compliant]
----

> type MyType = MyOtherType | null;

Reviewer: ```suggestion
type Foo = T | null;
```

> type MyType = any;

Reviewer: ```suggestion
type Bar = any;
```

> type MyType = MyOtherType & AnotherType;

Reviewer: ```suggestion
type Baz = T & U;
```
